designs/2019-relative-plugin-loading/README.md

```diff
@@ -0,0 +1,169 @@
+* ESLint loads plugins from the CWD of the process, which is typically somewhere in the end user's project. However, with the current setup the correct place to load plugins from is the ESLint invoker's project. As a result, if the ESLint invoker is not the end user (e.g. in `create-react-app` projects), the default plugin-loading location is incorrect. The ESLint invoker can change the CWD, but changing the CWD has some inconvenient side-effects:
+    * Only the `node_modules` folder in the CWD is ignored by default, so changing the CWD would cause some files to be unignored.
+    * Editor integrations usually won't know to change the CWD (since they might not be able to figure out who the "intended" ESLint invoker is for a given project), so they might fail to load the correct plugins.
+* Users often want to write custom project-specific rules and put them in the project's repository. However, plugins currently need to be in `node_modules`, because allowing other paths for plugins would introduce the possibility of name conflicts.
+
+## Detailed Design
+
+The design has three components:
+
+* Plugins are resolved relative to the config files that depend on them, rather than being loaded relative to the ESLint invoker's project.
+* Plugin name conflicts result in a fatal error.
+* Plugins in configs can be loaded by filepath.
+
+### Plugins are resolved relative to the config files that depend on them
+
+A config is said to "load" a given plugin name if either of the following are true:
+
+* The config includes the plugin name in a `plugins` list, e.g. `plugins: ["foo"]`
+* The config extends a shareable config from the plugin, e.g. `extends: ["plugin:foo/bar"]`
+
+To load a plugin from a config, the following steps are taken:
+
+* If the config loading the name `eslint-plugin-X` is itself a bundled config in plugin called `eslint-plugin-X`, then the load resolves to that plugin itself. (This is called a *self-load*.)
+* Otherwise, the plugin is loaded as a Node module relative to the location of the config file that loads the plugin's name.
+
+**Open question:** Should `extends: ["plugin:foo/bar"]` be considered a "load" of a plugin? (The "Open Questions" section at near the bottom of this RFC contains more details on how this would affect behavior. To understand the tradeoffs, it may be easier to read through this RFC first before considering the open question.)
+
+### Plugin name conflicts result in a fatal error
+
+If at least two different config files load the same plugin name, excluding self-loads, then a fatal config error is thrown.
```

Comment:
I have reconsidered about this in #13: https://github.com/eslint/rfcs/blob/cdf29fbae956803e897dcaadbf4a76c7ca0686fd/designs/2019-eslintrc-improvements/README.md#%EF%B8%8F-it-looks-plugins-up-from-the-location-where-the-config-file-is

In my updated thought, the option 4 of https://gist.github.com/not-an-aardvark/169bede8072c31a500e018ed7d6a8915 is the best. It allows users to use two config with a chosen version of the conflicted plugin without any special hack.

Reply:
To clarify, where would the plugin be resolved relative to, if it's loaded by both a parent and a child?

1. If it's resolved relative to the parent config, then we could decide whether to report conflicts for `plugin: ["foo"]` in child configs.
    1. If don't report conflicts in child configs (since only the version of the plugin relative to the parent would get loaded), then a user could encounter confusing schema errors if the configs start to depend on different versions of the same plugin. (For example, if one of the configs bumps the major version of a plugin, it will be incompatible with the user's version.)
    1. If we do still report conflicts in child configs, then a user would only be able to use two different configs depending on that plugin if one of  the plugins omits `plugins: ["foo"]`. (Otherwise, there would be a conflict since `plugins: ["foo"]` appears in both siblings.)
1. If it's resolved relative to the child config, then it could be confusing if the user separately installs `eslint-plugin-foo`, because the user's version of `eslint-plugin-foo` wouldn't get used.

Reply:
I'm considering it under #13.

- Every plugin is loaded. `eslint-config-foo` loads own plugin, `eslint-config-bar` loads own plugin, and `.eslintrc` loads own plugin.
- When `ConfigArray#extractConfig()` method merges the config array, it checks conflict. It traverses the config array from the last to the first, then
    - It adopts the first found plugin implementation.
    - If it found a plugin which has the same name, it compares the two. If the config file of the first found plugin "extends" the config file of the new found plugin, it just ignores the new found plugin. Otherwise, it throws conflict error.

For example, if `.eslintrc` doesn't have `plugins` property, `ConfigArray#extractConfig()` method adopts `eslint-config-bar`'s plugin and throws conflict error on `eslint-config-foo`'s plugin.
On the other hand, if `.eslintrc` has `plugins` property, `ConfigArray#extractConfig()` method adopts `.eslintrc`'s plugin and both `eslint-config-bar`'s plugin and `eslint-config-foo`'s plugin are ignored because those two is in ancestor.

Reply:
Thanks for clarifying. I think this is similar to "1i" from https://github.com/eslint/rfcs/pull/14#discussion_r260941864.

I think this wouldn't satisfy the robustness guarantee from [the overview gist](https://gist.github.com/not-an-aardvark/169bede8072c31a500e018ed7d6a8915).

```jsonc
// ./.eslintrc.json
{
    "plugins": ["foo"],
    "extends": ["bar"]
}
```

```jsonc
// ./package.json
{
    "dependencies": {
        "eslint-plugin-foo": "=1.0.0",
        "eslint-config-bar": "^2.0.0"
    }
}
```

```js
// ./node_modules/eslint-config-bar/index.js
module.exports = {
    plugins: ["foo"],
    rules: {
        "foo/baz": "error"
    }
};
```

If `eslint-plugin-foo` adds an option to the `foo/baz` rule that causes it to report fewer errors, then `eslint-config-bar` could upgrade `eslint-plugin-foo` and enable that option in a minor or patch release. However, this would cause schema errors for the end user's config, because an outdated version of `eslint-plugin-foo` would be used.

---

Separately from the robustness guarantee, I think it might be confusing that `plugins: ["foo"]` can silently change the behavior of shareable configs. If there was a rule schema error in a shareable config, it seems like it would be difficult to debug because the user could be using plugin versions that the shareable config doesn't support.

Reply:
I got it. My thought was about two shareable config with a plugin conflict, but I had not thought the end user pins the plugin which shareable config is using. hmm.